### ISSUE TITLE
[Suggestion] Automatic reconnect

### ISSUE THREAD

**Issue by DaFri-Nochiterov** [Genymobile/gnirehtet#47]
Could you please add function to the Android app or to the desktop one, so `gnirehtet` will be reconnected after was disconnected due to USB cable unplug (Once USB is plugged again I mean). So you don't need to restart desktop app in order to reconnect your device (I currently haven't found yet command to reconnect it from the device). 

Thank you for great app!

**Comment by rom1v** [comment]
The client attempts to reconnect automatically to the tunnel, hence to the relay server. Therefore, if the relay server is stopped and restarted, the client reconnects.

Unfortunately, unplugging the USB cable kills the `adb reverse` tunnel.

> So you don't need to restart desktop app in order to reconnect your device

In fact, restarting the tunnel is sufficient:

    adb reverse tcp:31416 tcp:31416

But I agree, this is far from ideal.

However, I don't want the relay server to poll the result of `adb devices` for detecting new devices.

Maybe I should expose a command `./gnirehtet tunnel` or something like that, to restart only the tunnel.

**Comment by theprogrammer67** [comment]
When the USB cable is disconnected, the fn poll_lop cycle (relay.rs) continues to work indefinitely.
So conceived?
I fixed it myself. The application is terminated with ExitCode <> 0  in this case

**Comment by rom1v** [comment]
@theprogrammer67 `relay` is the server, it accepts any number of devices: it must not be shutdown when one is disconnected.

**Comment by theprogrammer67** [comment]
but if everything is disconnectrf? clients.len() = 0 ?

**Comment by rom1v** [comment]
Then, it waits for new clients to connect (e.g. some `gnirehtet start`). It's in the same state as when it initially starts: waiting for clients…

**Comment by theprogrammer67** [comment]
his does not happen to me. After disconnect, the server does not automatically connect if you plug the cable (

**Comment by theprogrammer67** [comment]
I'm talking about the Windows program

**Comment by rom1v** [comment]
> does not automatically connect if you plug the cable

You need to (re)start the client (or at least restore the `adb reverse` tunnel), with `gnirehtet start`.

**Comment by theprogrammer67** [comment]
I inattentively read the documentation. Thank you )

**Comment by rom1v** [comment]
This may sound unnecessary complicated, but gnirehtet is really designed to be a client-server app. It has been developed to be used programmatically (in our use case, we start the server once, and start the clients when necessary).

I added some commands like `gnirehtet run` to simplify common use cases (start once on one device), but the many-to-one client-server architecture is still visible…

**Comment by autolainen** [comment]
Want to share how I solved problem with client reconnection after usb cable unplugged and then plugged back in:

```java
       ...
       ,RELAY_RECONNECT("relay-reconnect", CommandLineArguments.PARAM_NONE) {
            @Override
            String getDescription() {
                return "Start the relay server in reconnect mode.";
            }

            @Override
            void execute(CommandLineArguments args) throws Exception {

                new Thread(() -> {
                    while (true) {
                        try {
                            Socket s = new Socket("127.0.0.1", 5037);
                            s.getOutputStream().write("0012host:track-devices".getBytes(Charset.forName("ASCII")));

                            byte buf[] = new byte[5 * 1024];
                            int r;
                            while ((r = s.getInputStream().read(buf)) >= 0) {
                                if (r > 0) {
                                    try {
                                        execAdb(args.getSerial(), "reverse", "tcp:31416", "tcp:31416");
                                    } catch (Exception ex) {
                                        Log.e(TAG, "Cannot start client", ex);
                                    }
                                }
                            }
                        } catch (Exception e) {
                            try {
                                wait(30 * 1000);
                            } catch (Exception e1) {
                                // ignore
                            }
                        }
                    }
                }).start();

                relay();
            }
```
I added this is piece of code to Main.class so now I can start the server with command 'java -jar ginerhtet-server.jar relay-reconnect' and connect my mobile device by usb many times and it gets internet connection automatically. Hope it'll make your life easier!

**Comment by rom1v** [comment]
@autolainen Awesome solution! Thank you, great work! :+1:

To test in a shell:

    netcat 127.0.0.1 5037 <<< 0012host:track-devices

It could even be a command (say, `relay-auto`) that **starts** immediately the client for every connected device (not only enable the tunnel for the very same serial – which may be null –, but really starts all clients).

This implies to parse the output of the socket.

https://android.googlesource.com/platform/system/core/+/android-8.1.0_r9/adb/protocol.txt
https://android.googlesource.com/platform/system/core/+/android-8.1.0_r9/adb/SERVICES.TXT#27
https://android.googlesource.com/platform/system/core/+/android-8.1.0_r9/adb/transport.cpp#374

**Comment by rom1v** [comment]
@autolainen Thanks to your work, I implemented the feature on the Java version (I will also implement it in rust).

I added two commands:

 - `gnirehtet autostart` executes `gnirehtet start` on every new connected device;
 - `gnirehtet autorun` executes `gnirehtet autostart` in background and `gnirehtet relay`in foreground.

The branch is [`wip/autostart`](https://github.com/Genymobile/gnirehtet/commits/wip/autostart). Feedbacks welcome.

I'll merge it in `dev` once the Rust version is implemented.
